tls: flaky TestHandshakeConnectionCancelations/cancel_outgoing_connection

https://github.com/libp2p/go-libp2p/runs/7157818312?check_suite_focus=true

```
  === RUN   TestHandshakeConnectionCancelations/cancel_outgoing_connection
      transport_test.go:178: 
          	Error Trace:	D:\a\go-libp2p\go-libp2p\p2p\security\tls\transport_test.go:178
          	Error:      	An error is expected but got nil.
          	Test:       	TestHandshakeConnectionCancelations/cancel_outgoing_connection
  === RUN   TestHandshakeConnectionCancelations/cancel_incoming_connection
  --- FAIL: TestHandshakeConnectionCancelations (2.08s)
      --- FAIL: TestHandshakeConnectionCancelations/cancel_outgoing_connection (0.26s)
      --- PASS: TestHandshakeConnectionCancelations/cancel_incoming_connection (0.01s)

https://github.com/libp2p/go-libp2p/runs/7745508053?check_suite_focus=true

```
  === RUN   TestHandshakeConnectionCancelations
      transport_test.go:47: using a ECDSA key: QmViCTRxC6Hr755XJf9AVoZ7aHrmLE9SYsRijJhqQScXth
      transport_test.go:47: using a Ed25519 key: 12D3KooWNGdBE6geq4W7E9VA3EtnmNZUFzASHVFQdfjcH1EpTcCn
  === RUN   TestHandshakeConnectionCancelations/cancel_outgoing_connection
      transport_test.go:211: 
          	Error Trace:	/Users/runner/work/go-libp2p/go-libp2p/p2p/security/tls/transport_test.go:211
          	Error:      	Target error should be in err chain:
          	            	expected: "context canceled"
          	            	in chain: 
          	Test:       	TestHandshakeConnectionCancelations/cancel_outgoing_connection
  === RUN   TestHandshakeConnectionCancelations/cancel_incoming_connection
  --- FAIL: TestHandshakeConnectionCancelations (0.06s)
      --- FAIL: TestHandshakeConnectionCancelations/cancel_outgoing_connection (0.05s)
      --- PASS: TestHandshakeConnectionCancelations/cancel_incoming_connection (0.01s)

--- a/p2p/security/tls/transport_test.go
+++ b/p2p/security/tls/transport_test.go
@@ -188,7 +188,7 @@ func (c *delayedConn) Read(b []byte) (int, error) {
 	return c.Conn.Read(b)
 }
 
-func TestHandshakeConnectionCancelations(t *testing.T) {
+func TestHandshakeConnectionCancellations(t *testing.T) {
 	_, clientKey := createPeer(t)
 	serverID, serverKey := createPeer(t)
 
@@ -202,7 +202,13 @@ func TestHandshakeConnectionCancelations(t *testing.T) {
 
 		errChan := make(chan error)
 		go func() {
-			_, err := serverTransport.SecureInbound(context.Background(), serverInsecureConn, "")
+			conn, err := serverTransport.SecureInbound(context.Background(), serverInsecureConn, "")
+			// crypto/tls' context handling works by spinning up a separate Go routine that watches the context,
+			// and closes the underlying connection when that context is canceled.
+			// It is therefore not guaranteed (but very likely) that this happens _during_ the TLS handshake.
+			if err == nil {
+				_, err = conn.Read([]byte{0})
+			}
 			errChan <- err
 		}()
 		ctx, cancel := context.WithCancel(context.Background())

